scala/private/coverage_replacements_provider.bzl

```diff
@@ -0,0 +1,85 @@
+#
+# Coverage Replacements are a mapping of normal compiled artifacts to
+# instrumented compiled artifacts.
+#
+# The intention is that the final test runner inspects the test
+# classpath and replaces artifacts by any mappings found in the
+# `replacements` field.
+#
+# Rules producing replacement artifacts should _not_ link the
+# replacement files as any of the default outputs via DefaultInfo,
+# JavaInfo, etc. This way, actions producing the replacement artifacts
+# will be executed on an as needed basis.
+#
+# Coverage Replacements use a provider and a helper aspect to
+# aggregate replacements files across the dependency graph.
+#
+# Under the hood, two providers are needed because Bazel doesn't allow
+# duplicate providers.
+#
+
+_CoverageReplacements = provider(
+    fields = {
+        "replacements": "hash of files to swap out",
+    },
+)
+
+_CombinedCoverageReplacements = provider(
+    fields = {
+        "replacements": "hash of files to swap out",
+    },
+)
+
+# the attributes used to form the dependency graph that we'll fold
+# over for our aggregation
+_dependency_attributes = [
+    "deps",
+    "exports",
+]
+
+def _combine(*entriess, base = {}):
```

Comment:
It's just because it's a 2d array of entry items (or a 1d array of entries). In the comprehension I have:
```python
[
    entry[_CoverageReplacements].replacements
    for entries in entriess
    for entry in entries
    if _CoverageReplacements in entry
]
```